### DISCUSSION TITLE
Enablling LSP to work with cpp2/cppfront

### DISCUSSION THREAD

**Discussion by APokorny** [hsutter/cppfront#762] [Q&A]
Is there already a way to get a working LSP to provide  at least minimal completion hints for a cpp2 file?
I thought maybe running cppfront  on the buffer content and then sending that to clangd.. 

Or even better is someone already cooking a cppfront in clang?

**Comment by vanceism7** [comment]
This would be really cool. To me, having an LSP server instantly legitimizes a language, +10! 

I've never implemented an LSP server before but I'm interested in trying to tackle something like this. I've started researching to see what it takes

**Reply by MaxSagebaum** [reply]
I thought about this a bit, since this was also a direction I wanted to try.

Here, just a few thoughts.
 - I would not recommend using clang for this. 
 - The language server should share the current code of cppfront.
 - In the end, a compilation of cppfront should give `cppfront` and `cppfrontd`.

I would be happy to help. I also have not implementing a language server before. If you have something or want to talk just contact me.

**Reply by vanceism7** [reply]
Yea I agree. I've seen some LSP projects written before where the LSP server used a different (or rolled their own) tool for parsing and error reporting than the compiler itself, the result seems bad. The compiler has to be the source of truth or you won't have a consistent experience. cppfront needs to power the server or this won't work properly.

I don't know if the server necessarily needs to be a compilation artifact of cppfront, but the server should be calling into the cppfront compiler to get whatever info it needs. In the worst case, cppfront might just require a few extra functions and/or CLI args to report out on its results. 

I don't think we even need to interface with a c++ parser to make this work (like clangd or anything of the sort), I believe cppfront already knows how to parse c++ on its own (but Im not positive on this).

@MaxSagebaum - if you're curious, I found this tutorial from Microsoft which shows how to build a simple LSP server. It can probably be used as a starting point to write our own server. Ive been reading through it but haven't actually done anything yet. But I was thinking of copying it and pushing up a repo to get started

https://code.visualstudio.com/api/language-extensions/language-server-extension-guide

**Reply by hsutter** [reply]
> I believe cppfront already knows how to parse c++ on its own (but Im not positive on this).

Cppfront parses Cpp2 code only. It allows mixed Cpp1/Cpp2 source files, but passes through the Cpp1 code without parsing it (I do just the minimum necessary to know when the Cpp1 code ends, including brace matching which requires the lexical parse of character and string literals since those could contain braces the brace-matching shouldn't count).

**Reply by vanceism7** [reply]
Ah I see, thanks for confirming that for us. That's an interesting situation then. I'm usually pretty naive when it comes to all this stuff, but it sounds like our options are:

1. The lsp server to have a reference to both a cpp compiler and cppfront so it can gather info from both sets of data
2. We always compile the cpp2 to cpp1 and then use a cpp1 compiler to analyze the data
3. We just support pure cpp2 files at first.

Option 2 sounds like it might be the simplest and least invasive approach (since it hopefully requires 0 changes to cppfront) and also reaches the strongest feature set. This is basically how the other tooling is working too, like gdb debugging and all that. Actually this still requires having a reference to a cpp1 compiler and cppfront. Also, I think this somewhat throws us back into "lets use clangd" territory, but maybe this isn't a bad thing. It seems like one of the big selling point of cppfront this whole time has been getting all of our tools "for free"

**Reply by MaxSagebaum** [reply]
Thanks for the link. I will have look during the weekend.

I think option 2. will work but will be the most clunky one. Cppfront adds some extra stuff in the code, so the compiler might have a hard time to detect e.g. the member of a structure. Nevertheless, for a fast result it might be the best option. 

As a start, I think option 3 would be the way to got. The cpp1 parts of the file could be ignored at first. If this is stable the server could be extended to talk with a language server for cpp1 and make a request also there. This information can then be added to the data from cppfront. For a full language server the starting point would probably be sema.h. This would give you all compile errors.

**Reply by APokorny** [reply]
Option 4: Also parse cpp

I mean how hard can it be.

**Reply by DyXel** [reply]
![parse_cpp_clueless](https://github.com/user-attachments/assets/c360f6e1-666d-4c76-b720-ffa3edef31f9)

**Reply by JohelEGP** [reply]
See [Design note: Unambiguous parsing's \(\*\*\*\) Parsing issues in today's C++](https://github.com/hsutter/cppfront/wiki/Design-note:-Unambiguous-parsing#-parsing-issues-in-todays-c).

**Reply by vanceism7** [reply]
I once upon a time tried to write a language that transpiles to C, and C also has some of the same issues where the meaning of the parse depends on the semantic analysis. I don't think it's as difficult as the examples Herb mentioned in his design notes, but it's still difficult! I don't think that (parsing c++) is a path we want to go down lol

**Reply by MaxSagebaum** [reply]
I had a look at the specification. It seems to be simple enough, that a self implemented solution could be feasible. One question would be which language to choose for the server. Cpp2 or Cpp1? You proposed one other interesting idea that Cppfront should output all information about the sourcefile in a json file. The server could then use this information. A short search did not provide any insights if such a framework exists. I also did not found a general framework for implementing a language server. 

Herb specifically did not want to use any frameworks when implementing cppfront. Maybe this is also a way for the language server?

So what would you favorite way be? 
 - cppfront -> details in json file -> server that gets information from the json file?
 - cppfront -> internal data structure -> server that works on the internal data structure?

**Reply by vanceism7** [reply]
Hmm, the choice of which language to use was one I was thinking about too. I think using c++ is the fun option, and cpp2 is the even funner option, but I don't know that it's really necessary. The easiest option would be to just use typescript - then it's just a matter of implementing the couple of LSP feature functions we need and we're done. All of this is already stubbed out in that link I shared/the example repo they provide; I copied that example and began modifying it to implement the cpp2 functionality I shared in my comment below.

Not that I'm necessarily attached to the idea of using typescript; perhaps there are reasons why c++ is the better option that just I haven't thought about. 

But I think regarding your question of how we work with cppfront: whether to use json or internal structure depends on which language we pick for the language server. I think we can only use internal structure if we choose cpp/cpp2, otherwise we must use json (or some other serialized format) to interface with the data. Using json gives a nice interface that allows us more flexibility of what language we want to implement in

So yea, those are my initial thoughts on it. What do you think?

**Reply by MaxSagebaum** [reply]
From what Herb wrote below, I think the currently minimal invasive approach would be to have an to_json.h file. So that this can actually be merged into cppfront. Thinking about it, it has 3 advatages:
 - Writing out first simple json items should be straight forward.
 - Writing the server in a separate project removes cppfront contraints and libraries can more easily be used. E.g. a json parser.
 - It lets you (us) gather experience on what kind of data is actually required.
 
Migrating the server to cpp1/cpp2 should then be simpler, since we know what kind of data we need. The `to_json.h` could then be changed into an `to_lsp.h`, which produces the required data structure directly.

Would this be feasible?

**Reply by vanceism7** [reply]
Yea, if we think implementing the language server in c++ is a goal worth pursuing, then this is probably the way to go. Really I think we need to aggregate this data regardless of what path we choose. As I've been playing with the cppfront code, I looked at the `debug_print` code that gets triggered when you use the `-debug` flag and you can see in that function that you basically have access to all of the compilers diagnostics right there. (Namely, `source`, `tokens`, `parser`, and `sema`)

https://github.com/hsutter/cppfront/blob/68b716abe3e4b623e875d68a2b9b3f0efcd39b45/source/to_cpp1.h#L7176-L7195

So I think the first step would just be to aggregate the data from these fields into something that is more easily consumable from another program. I think particularly, the names of all symbols defined in the file, and all of the errors definitions (which can be found in `sema.errors`). I've been poking at this a bit but haven't made too much progress yet

**Reply by MaxSagebaum** [reply]
This sound like a good first step. If you need some help just contact me.

**Comment by hsutter** [comment]
Thanks for the idea!

Info in case it helps... cppfront's linear structure is this, where each file depends only on the one immediately above it:

```cpp
common.h
io.h                    Cpp1/Cpp2 source file i/o
lex.h                   Cpp2 tokenizer
parse.h                 Cpp2 parser
reflect.h2              Cpp2 reflection API (with cpp2regex.h)
sema.h                  Cpp2 semantic analysis
to_cpp1.h               Cpp2->Cpp1 lowering
cppfront.cpp            Driver
```

Each subset is intended to be standalone and reusable, so:

- If you want a Cpp2 parser (only), you could try reusing `parse.h` and ignore what's below it.
- If you want the reflection API too, you could go one step down and try `reflect.h2` (use the generated `reflect.h` if you're using it from Cpp1 code).
- If you want to actually generate parts of the Cpp1 code to show/analyze/pretty-format/etc., you could go down to `to_cpp1.h` (check out `print_to_string` which lowers to a `std::string`).
- If you want the whole Cpp1 code, you could run the whole compiler and analyze/pretty-format/etc. the entire generated output `.cpp` file.

If you see that it "almost" works and you need something more that's not there, and it's super easy to add (especially if it's a lightweight PR, since code is appreciated) it could fit in #1287's Priority 1. Disclaimer: I don't currently have capacity to do larger-scale refactoring/reorganization or invasive changes though.

HTH!

**Comment by vanceism7** [comment]
I cobbled this together today

![image](https://github.com/user-attachments/assets/78e2804a-acfd-494c-a02c-74ca9b9761b6)

I had to (or rather, wanted to) add a flag to cppfront to output its error results as json, just so it was easier to parse the error reports without too much fuss. It's probably bad... but perhaps I'll open a PR

**Reply by hsutter** [reply]
Nice!

Re JSON: I don't have an issue with that, but just curious what other compilers do -- do they have switches for emitting errors as JSON, or do they do something else?

**Reply by vanceism7** [reply]
My experience with compilers is super limited, but IIRC, I believe at-least the [Purescript](https://github.com/purescript/purescript) guys did this with their compiler to make it easier to integrate tooling.

Actually, as I'm thinking about it more, I think they didn't just emit errors as json, but a whole slew of compiler information was written as json. If that's the case, the PR I just opened may be too limited in scope, but maybe its still a good starting point

**Reply by zaucy** [reply]
While not all compilers here are a few 'compiler adjacent' tools that output json stream (json objects separated by newlines). I'm sure there's plenty more but these are the ones off the top of my head

* `bazel query` - `--output=streamed_jsonproto` (json to stdout)
* `bazel build` - `--build_event_json_file=/path/to/json` (json to file)
* `rustc` (compiler!) - `--error-format=json` (json to stderr)
* `fastbuild` - `-report=json` (json to stdout/stderr)

**Reply by JVApen** [reply]
> Nice!
> 
> Re JSON: I don't have an issue with that, but just curious what other compilers do -- do they have switches for emitting errors as JSON, or do they do something else?

Clang has https://clang.llvm.org/docs/UsersManual.html#formatting-of-diagnostics
I believe they also support outputting SARIF, which I believe to be  structured JSON

**Comment by vanceism7** [comment]
I've been working on this language server for a little bit now, and have managed to implement `goto definition` and `autocompletion`.

Figured I'd paste it here in case anyone wants to check it out. It's all currently in typescript, I've been using it as a test-bed to inform the types of info I need out of cppfront.

https://github.com/vanceism7/ls-cpp2

**Reply by MaxSagebaum** [reply]
Nice, I will have a look tomorrow.

**Reply by MaxSagebaum** [reply]
Ok, the tomorrow took  a little bit longer. Sorry for that. 

I had a look at the current state and it is quite promising. It worked nearly out of the box, but I had a few issues:
 - the `unuri` method removed one slash to many. The server tried to run cppfront with files like `home/msagebaum/cppfront/source/reflect.h2-diagnostrics.json` for which the path was not available.
 - I tried the file `reflect.h2` from cppfront (the one from the lsp-branch) and cpp front crashes here. The json file is the empty, which lead to the error: `Error parsing json  SyntaxError: Unexpected end of JSON input`

The completion worked and also the goto. I noticed one problem which mainly comes from the way how cppfront handles the parsing. If you are in a function and are typing, you get the error reports from cppfront, like there is a missing `;` because you have not finished the line yet. Since cppfront directly reports the error and does not finish the parsing, the symbol table is incomplete. You get the completion from everything defined beforehand  but nothing from the function itself and afterwards.

With errors present:
![image](https://github.com/user-attachments/assets/537e11f3-aa92-40c0-933d-04e171662c24)

If no errors are present:
![image](https://github.com/user-attachments/assets/6277431e-d511-45a2-8741-67f68eb19073)

This is mostly how cppfront handles the parsing. Maybe it could be circumvented by the following approach:
 - Run cppfront on the file.
 - If the json contains no errors, update the symbols for the file.
 - If the json contains errors, do not update the symbols fro the file and use the symbols from the last valid run. Add the errors to the report.

If you want I could investigate the crash of cppfront on `reflect.h2`. 

I could also have try at implementing the above "algorithm" but I have never written any JavaScript. Do you have any pointers?

**Reply by vanceism7** [reply]
> Ok, the tomorrow took a little bit longer. Sorry for that.

😅 No problem haha, sometimes tomorrow can be a pesky thing lol.

> If you want I could investigate the crash of cppfront on reflect.h2.

Ah gotcha! Thanks for finding this bug. I haven't tried the language server on anything more than my toy cpp2 files so I'm sure the crash was bound to happen. Feel free to dig deeper if you want, but otherwise I'm happy to give it a more proper test myself as well.

> I could also have try at implementing the above "algorithm" but I have never written any JavaScript. Do you have any pointers?

Actually, the generation of the diagnostics file happens in cppfront itself. Currently the entire json file is entirely overwritten each time cppfront compiles the source, so we'll need to get more sophisticated if we want to update the errors but not the symbol table. But yea, it's implemented in cppfront so it'd be implemented in c++ rather than JavaScript.

I'm open to all the help I can get! I started the language server more to "get the ball rolling" than anything. I was hoping if it caught on, people much smarter than I would help it to become more concrete 😄

**Reply by MaxSagebaum** [reply]
I had a look at the segfault. It was caused by an anonymous function declaration. I created a pull request with the fix. Now the language server can parse reflect.h2. 🎉🎉🎉

A completion in a random code location looks now like:
![image](https://github.com/user-attachments/assets/e888fe51-fe84-4f16-aede-c6927a097500)


> Actually, the generation of the diagnostics file happens in cppfront itself. Currently the entire json file is entirely overwritten each time cppfront compiles the source, so we'll need to get more sophisticated if we want to update the errors but not the symbol table. But yea, it's implemented in cppfront so it'd be implemented in c++ rather than JavaScript.

Currently, a better option is to handle this in the language server. The server has to read the json file anyway. It can have a quick look at the content and perform the caching if no errors are there. If there are errors, it takes the symbols from the last cache and the current errors. What do you think?

**Reply by vanceism7** [reply]
Yea that's a good idea. Previously, the server would parse the json into some `Diagnostics` object type and then say something like
`currentDiagnostics = newDiagnostics`

We basically just need to do something more like

```
currentDiagnostics = currentDiagnostics.mergeWith(newDiagnostics)
```

Where `mergeWith` is smarter about choosing how to update the data

**Reply by MaxSagebaum** [reply]
Ok, sounds good. I never programmed in type script. So you are probably faster in doing it? Or should I have a look?

**Reply by vanceism7** [reply]
It's probably faster for me to do it. I'll see if I can give it a look tomorrow. But feel free to check it out if you want. Typescript is super easy, I'm sure you'd get the hang of it quickly haha

**Reply by vanceism7** [reply]
Ok - I pushed an update that implements the merging logic, so it should now do a better job of not losing symbols in the symbol table when errors occur in the document

**Reply by MaxSagebaum** [reply]
It works now quite well now, even on `reflect.h2`. There seems a subtile bug, where the server loses the definitions again and you get only the text completion from VSC. I do not know if this is because of the server or VCS. I could not find a procedure to reproduce it consistently.

What is your plan for next steps? From my point of view two things are missing:
 - hints/completion for function arguments. Like name of the argument, different versions ![image](https://github.com/user-attachments/assets/ea59363c-e9cd-4d6d-ab66-9a68689890b7)
 - completion for members of structures 
![image](https://github.com/user-attachments/assets/efafd093-17ff-489c-af15-470e3551f920)

This probably only concerns the generated json file, or are these special commands to the server. I will have some time in February where I could have look.

**Reply by vanceism7** [reply]
> What is your plan for next steps? From my point of view two things are missing:
> - hints/completion for function arguments. Like name of the argument, different versions
>   
> ...
> This probably only concerns the generated json file, or are these special commands to the server. 

I think it's both for this one. When the type of symbol is something more sophisticated than a variable, e.g: a function or class, we need to augment the symbols in the json diagnostics with that information. During the check for the symbol type, I think `sema` may have more info available, such as function argument type info.

Once we have this info, we need to augment the server to read it in and include it in the language servers completion hint details. This is likely trivial compared to collecting the info

> - completion for members of structures

Specifically regarding this one, I was looking into it but from what I was seeing, cppfront doesn't seem to capture any symbol information for class/struct members variables. I opened an issue for that:
https://github.com/hsutter/cppfront/issues/1323

I don't know if it was a bug or intentional - it could be that the intention was just to defer those symbols to the cpp compiler.

> There seems a subtile bug, where the server loses the definitions again and you get only the text completion from VSC.

I noticed a similar weird bug but as you said, it wasn't consistent. It may have to do with the fact that cppfront is generating a new diagnostic file every single key press - it could be working out the file system with too much io. Adding a debouncer to the compiler that waits for a 1-2 seconds of inactivity before recompiling the file might help here.

The other big thing is I wanted to start capturing symbols from the cpp compiler, but this info isn't published in the sarif output of the compilers. I think its output to the .obj files or something of that sort, so that part will take a bit more time to implement.

**Comment by JVApen** [comment]
How do you see interactions between Cpp1 LSP (MS C/C++ or clangd) and Cpp2 LSP?
I was under the impression that a Cpp1 LSP could perfectly work with the converted sources. At least for things like auto-complete, that would be sufficient.

However, if I start a rename operation in my Cpp1 LSP, it would suggest renaming the callers in the preprocessed code, not the Cpp2 code. As such, after the rename, everything seems to be working until the generated file gets regenerated. At which point, your build is broken.

Vice versa, I see auto-complete in Cpp2 only usable if it can give suggestions of functions defined in pure Cpp1. A rename started from Cpp2 code that is used in Cpp1 code has the exact same problem.

Note: this problem holds for all C++ successors, though Cpp2 seems to be the first to reach this point. Actually, it holds for all C++ code generators, for example lex/yacc. For any successor, you expect that over some time it covers sufficient code that this problem can no longer be ignored.

In order to solve this, you either need 1 LSP that understands both Cpp1 and Cpp2 or you should find a good way to have them collaborate.

Semantic code rewrites (like clang-tidy) might even result in a harder problem to solve. Having multiple successors in a single codebase will even be more challenging.

**Reply by vanceism7** [reply]
> How do you see interactions between Cpp1 LSP (MS C/C++ or clangd) and Cpp2 LSP? 
...
> In order to solve this, you either need 1 LSP that understands both Cpp1 and Cpp2 ...

This is basically the answer, at least in my mind. From what I've learned so far working on this language server, there's really no way to make the server fully functional without incorporating c++ diagnostics. Trying to interface two language servers together also seems like tackling the problem at the wrong abstraction level.

I did initially explore this problem from the angle of trying to extend something like clangd to understand cpp2, it intuitively feels like it wouldn't be that hard, but after some discussion with the llvm folks, it appeared that this approach requires non-trivial work going directly into clang itself. That's bad because it's hard *and* because a cpp2 language server shouldn't be dependent on a specific compiler.

So in the end, I think the right approach is to pull diagnostics from cppfront, and then compile the file and pull diagnostics from the users chosen c++ compiler. As far as I've read, all of the major c++ compilers do have the ability to output diagnostics in a way that can be consumed by external tools - I assume this is how the other c++ language servers work. The cpp2 server will also need to manage mappings from cpp1 code to cpp2 so we can consume the cpp1 diagnostics correctly. Generating and consuming the cpp1 data is the next major piece I'll be looking at soon.

**Reply by JVApen** [reply]
I saw that discussion and responded in it, though it felt like it stopped without a clear track forward.

Consuming the Cpp1 diagnostic and mapping it back to Cpp2 seems indeed like the most reasonable approach. I'm wondering if this is something you can do for most of the LSP calls.

For example: syntax highlighting of Cpp1 code,except the location, can stay unmodified, as soon as it reaches Cpp2, you include your own code. Fixing the offsets seems quite possible.

For refactorings (rename, tidy, formatting), you most likely need to translate changes to the generated code and map it back to Cpp2. Especially decomposing the changes will be challenging here.

For auto-complete you might be able to first map your cursor onto the generated code and pass it along after which you can convert back. The only annoying thing here (and any LSP) is that you should be able to deal with incomplete code. In case of Cpp2, that would imply generating broken Cpp1 code first.
The same will most likely hold for hover.

As far as I'm aware, there are only 1-3 calls where the full source code is send to the server, all others use a location in the previously known code. So if you want to do preprocessing there, it should be possible. For all other requests, you mainly should be able to do the mapping of location towards the Cpp1 LSP and do the reverse mapping and code rewrite on responses.

The background indexing is most likely going to be the most challenging as it will find preprocessed files without going through your wrapper.

The more I think about this, the more feasible it seems to write a wrapper around the whole LSP and let the Cpp1 LSP handle the majority of the work, even for Cpp2 syntax. Although getting all details right is going to be challenging. This should in theory be LSP unaware, though every LSP has its own extensions.
It still sounds more complex than teaching Clang to parse Cpp2, though even in that case you need to somehow give responses in Cpp2 syntax.

If this works out, a more integrated plugin approach can be introduced in clangd and others LSPs, which might also work for lex/yacc and other code generators. Maybe some generic meta-data can simplify it even more if we have more of them. Though that's something to only worry about once something works.

**Reply by JVApen** [reply]
I'm not convinced that you should be using every compiler as backend for your LSP. Even Microsoft uses a separate parser for intellisense than the compiler. It also doesn't change when you would compile with GCC for Linux or Clang for any platform.

So it seems reasonable to focus on a single LSP first. If you don't want to wrap completely, you can hack into your own clone of clangd. All requests come in here: https://github.com/llvm/llvm-project/blob/229abcd459dc365201185aa3988b9f8ae455de76/clang-tools-extra/clangd/ClangdLSPServer.cpp#L1659

I wish I had the time to actually assist in coding.

**Reply by vanceism7** [reply]
I've been thinking about what you said here. I think you're right. I just realized that going down the path of integrating diagnostics from c++ compilers is essentially writing a language server for c++. It seems like the easier approach would be to clone a c++ language server and augment it with cpp2, rather than the reverse, otherwise we're duplicating _a lot_ of effort!

Thanks for the reference to the clangd code. I was poking around on that project a bit but hadn't really made sense of it yet. The link you gave provides a nice starting point.

**Reply by vanceism7** [reply]
I have some good news to report regarding the concern of building for multiple compilers - it actually looks like Microsoft has been working on a diagnostics standard that will allow for easier consumption from multiple compilers (as well as other dev tooling). The search took me a while, but whilst trying to figure out how to get compiler results in json from the major compilers (msvc, clang, and gcc), I discovered that the json-like format they all share in common is something called [SARIF](https://sarifweb.azurewebsites.net/). 

Even though its duplicate work, creating a language server that pulls its diagnostic results directly from the compiler could be beneficial to give more accurate error messages. (I've already noticed both clangd and microsoft's c++ ls both have issues where they report errors that aren't really there) As a bonus, it also makes the language server more flexible since it can work with whatever compiler the user chooses to use. So yea, perhaps even though it's duplicating effort in some sense, it could be worth it since sarif standardizes results (hopefully meaning we only need to duplicate the effort once, and then it works for all sarif compliant compilers afterwards). Gonna experiment with stuff more

I just noticed you mentioned SARIF in your other response up above. Just for general info for anyone reading, each compiler can output to sarif as follows

`clang++ -fdiagnostics-format=sarif file.cpp`
`cl.exe /EHsc /experimental:log .\sarif-output-name .\file.cpp`
`g++ -fdiagnostics-format=sarif-file file.cpp`

**Comment by vanceism7** [comment]
A new update. All of the basic cpp2 diagnostics themselves are working, and I've just managed today to augment them with the standard c++ compiler errors now too. Although the diagnostics from the c++ compiler are a little off... At this point, we need a better mapping between the generated c++ code and the cpp2 code in order to show more accurate error locations/reports. 

A big bulk of the work is taken care of automatically thanks to the `#line` pragmas, but the difference in syntax causes the column number to be off a bit. I'm not sure if we're catching line/column numbers for symbols in cpp2 code _and_ c++ code (or atleast just the columns), but I think if cppfront was capturing this information, it'd enable us to get a pretty flawless translation between c++ and cpp2 code for error reporting.

![image](https://github.com/user-attachments/assets/ecd8871f-7fbe-4869-934c-0aad1f72f287)
